pkg/util/push/push.go

```diff
@@ -84,6 +94,33 @@ func handler(maxRecvMsgSize int,
 				logger = log.WithSourceIPs(source, logger)
 			}
 		}
+
+		// TODO: move this to middleware or better place
+		cluster, replica := r.Header.Get(ClusterHeader), r.Header.Get(ReplicaHeader)
```

Comment:
Instead of parsing this here, where we still don't even know whether we want to check replicas (if HA dedup is disabled, or it's a different usage), we should just replace `replicaChecker` with:

```go
shortcut func(ctx context.Context, w http.ResponseWriter, req *http.Request) ( finished bool)
```

This function should return `finished` only if it knows cluster/replica and has taken a decision to discard the request responding:

```go
w.WriteHeader(http.StatusAccepted)
w.Header().Set(IsSecondaryRWReplicaHeader, "true")
```

so the function we pass should look like this (writing in the browser, this may fail, but you get the idea):

```go
func(ctx context.Context, w http.ResponseWriter, req *http.Request) (finished bool) {
	cluster, replica := r.Header.Get(ClusterHeader), r.Header.Get(ReplicaHeader)
	if cluster == "" || replica == "" {
		// Nothing to shortcut.
		return false
	}

	userID, err := tenant.TenantID(ctx)
	if err != nil {
		// Someone will handle this, we can't shortcut.
		return false
	}
	
	err = replicaChecker(ctx, userID, cluster, replica, time.Now())
	// NOTE: we should use `errors.As` here, because we're checking the type. That's why we don't need the hacky Is() implementation.
	if errors.As(err, &ReplicasNotMatchError{}) {
		// Replica does not match, we can shortcut.
		w.WriteHeader(http.StatusAccepted)
		w.Header().Set(IsSecondaryRWReplicaHeader, "true")		
		return true
	}
	
	// Can't shortcut, process normally.
	return false
}
```
		

If something fails (no user? wrong header, etc. we should not stop the request, we already have mechanisms for failing properly there).